pkg/networkservice/common/begin/event_factory.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package begin
+
+import (
+	"context"
+
+	"github.com/edwarnicke/serialize"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/postpone"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type connectionState int
+
+const (
+	zero connectionState = iota + 1
+	established
+	closed
+)
+
+var _ connectionState = zero
+
+// EventFactory - allows firing off a Request or Close event from midchain
+type EventFactory interface {
+	Request(opts ...Option) <-chan error
+	Close(opts ...Option) <-chan error
+}
+
+type eventFactoryClient struct {
+	state      connectionState
+	executor   serialize.Executor
+	ctxFunc    func() (context.Context, context.CancelFunc)
+	request    *networkservice.NetworkServiceRequest
+	opts       []grpc.CallOption
+	client     networkservice.NetworkServiceClient
+	afterClose func()
+}
+
+func newEventFactoryClient(ctx context.Context, afterClose func(), opts ...grpc.CallOption) *eventFactoryClient {
+	f := &eventFactoryClient{
+		client: next.Client(ctx),
+		opts:   opts,
+	}
+	ctxFunc := postpone.Context(ctx)
+	f.ctxFunc = func() (context.Context, context.CancelFunc) {
+		eventCtx, cancel := ctxFunc()
+		return withEventFactory(eventCtx, f), cancel
+	}
+
+	f.afterClose = func() {
+		f.state = closed
+		if afterClose != nil {
+			afterClose()
+		}
+	}
+	return f
+}
+
+func (f *eventFactoryClient) Request(opts ...Option) <-chan error {
```

Comment:
During local tests running from pkg `begin` I found that `eventFactoryClient.Request`/`eventFactoryClient.Close`  and `eventFactoryServer.Reqest`/`eventFactoryServer.Close` never call. 

Is it expected?

Reply:
My intention was to let them be covered by refresh and timeout...